[export] Implement cpp deserializer.

zhxchen17 opened the pull request:
Differential Revision: D63206258

This diff introduces a mechanism to generate a json-compatible deserializer in cpp using nlohmann json (already being used by AOTI). 

Why we need this? Because there will be a lot of cases where people don't want to use Python to load the graph (e.g. cpp runtime), and instead they can use this header to deserialize the JSON graph.

Every time we call update_schema.py to update the schema, the header will be auto generated and included into the source files.

seemethere reviewed (changes requested):
You should add the generated file to the list of generated files here:

https://github.com/pytorch/pytorch/blob/main/.gitattributes

Then the check should pass since it ignores generated files.

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -770,13 +770,13 @@ def serialize_input(
             # self.is_sym_int_arg(arg) being true
             return Argument.create(as_sym_int=SymIntArgument.create(as_name=str(arg)))
-        elif isinstance(arg, bool):
+        elif type(arg) is bool:
```
I think it's for keeping consistency with isinstance(arg, int). no difference I know of, but instance(arg, bool) might have surprises we don't know, like int.

zhxchen17 commented on torch/_C/_aoti.pyi:
```diff
@@ -16,6 +16,15 @@ def alloc_tensor_by_stealing_from_void_ptr(
     handle: c_void_p,
 ) -> Tensor: ...
 
+class CppExportedProgram: ...
```
oki

zhxchen17 committed:
[export] Implement cpp deserializer. (#136398)

Summary:

This diff introduces a mechanism to generate a json-compatible deserializer in cpp using nlohmann json (already being used by AOTI).

Why we need this? Because there will be a lot of cases where people don't want to use Python to load the graph (e.g. cpp runtime), and instead they can use this header to deserialize the JSON graph.

Every time we call update_schema.py to update the schema, the header will be auto generated and included into the source files.

Test Plan:
I added a new variant of tests called "CppSerdes" which contains hundreds of new tests.

buck test mode/opt caffe2/test:test_export -- -r cpp_serdes

Reviewed By: angelayi

Differential Revision: D63206258

zhxchen17 commented:
@pytorchbot merge -i
